### PR TITLE
implement inline command for dependency: Jump to definition

### PR DESCRIPTION
**Pull request by Melody618**
this PR implements an inline command, jump to definition, for dependency node in sidebar. If click, it will open the pom file where the dependency is imported.

https://user-images.githubusercontent.com/25446374/126265383-c7a6b756-9207-4814-8eeb-b5db7929ff5e.mp4

### CODE REVIEW

File: package.nls.zh.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "contributes.commands.maven.view.flat": "切换到扁平视图",
     "contributes.commands.maven.project.addDependency": "添加依赖…",
     "contributes.commands.maven.project.showDependencies": "显示所有依赖",
+    "contributes.commands.maven.project.jumpToDefinition": "打开定义位置",
```
Comment (Eskibear) [inline package.nls.zh.json:18]: image

File: package.nls.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
     "contributes.commands.maven.project.addDependency": "Add a dependency...",
     "contributes.commands.maven.project.showDependencies": "Show dependencies",
     "contributes.commands.maven.project.excludeDependency": "Exclude dependency",
+    "contributes.commands.maven.project.jumpToDefinition": "Jump to Definition",
```
Comment (Eskibear) [inline package.nls.json:19]: "Go to Definition" for consistency

File: package.json [hunk 3]
```diff
@@ -58,6 +58,7 @@
     "onCommand:maven.plugin.execute",
     "onCommand:maven.project.addDependency",
     "onCommand:maven.project.excludeDependency",
+    "onCommand:maven.project.jumpToDefinition",
```
Comment (Eskibear) [inline package.json:61]: Is it necessary here? In which case do you activate extension by running this command?

File: src/handlers/jumpToDefinitionHandler.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import * as path from "path";
+import * as vscode from "vscode";
+import { Dependency } from "../explorer/model/Dependency";
+import { getMavenLocalRepository } from "../utils/contextUtils";
+import { UserError } from "../utils/errorUtils";
+import { ElementNode, getNodesByTag, XmlTagName } from "../utils/lexerUtils";
+
+export async function jumpToDefinitionHandler(node?: Dependency): Promise<void> {
+    if (node === undefined) {
+        throw new Error("Only Dependency can jump to definition.");
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:14]: grammar problem.  It's "you" jump to the defnition of a dependency node. 

"No dependency node specified." would be just fine.

File: src/handlers/jumpToDefinitionHandler.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+async function jumpToDefinition(pomPath: string, aid: string): Promise<void> {
+    const contentBuf: Buffer = await fse.readFile(pomPath);
+    const projectNodes: ElementNode[] = getNodesByTag(contentBuf.toString(), XmlTagName.Project);
+    if (projectNodes === undefined || projectNodes.length !== 1) {
+        throw new UserError("Only support POM file with single <project> node.");
+    }
+
+    const projectNode: ElementNode = projectNodes[0];
+    const dependenciesNode: ElementNode | undefined = projectNode.children?.find(node => node.tag === XmlTagName.Dependencies);
+    const dependencyNode: ElementNode | undefined = dependenciesNode?.children && dependenciesNode?.children.find(node =>
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:36]: `dependenciesNode?.children && dependenciesNode?.children.find()` can be shorten to `dependenciesNode?.children?.find`

File: src/handlers/jumpToDefinitionHandler.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        node.children && node.children[1].tag === XmlTagName.ArtifactId && node.children[1].text === aid
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:37]: 1. `children[1]` is depending on order of child items?
2. what if two artifacts with same aid?  e.g. gid1:aid vs gid2:aid

File: src/handlers/jumpToDefinitionHandler.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    if (node.parent === undefined) {
+        selectedPath = node.projectPomPath;
+    } else {
+        const parent: Dependency = <Dependency> node.parent;
+        selectedPath = localPomPath(parent.groupId, parent.artifactId, parent.version);
+    }
+    await jumpToDefinition(selectedPath, node.artifactId);
+}
+
+async function jumpToDefinition(pomPath: string, aid: string): Promise<void> {
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:27]: pomPath + aid are not sufficient to specify the location, AFAIK.

File: src/handlers/jumpToDefinitionHandler.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    const projectNode: ElementNode = projectNodes[0];
+    const dependenciesNode: ElementNode | undefined = projectNode.children?.find(node => node.tag === XmlTagName.Dependencies);
+    const dependencyNode: ElementNode | undefined = dependenciesNode?.children && dependenciesNode?.children.find(node =>
+        node.children && node.children[1].tag === XmlTagName.ArtifactId && node.children[1].text === aid
+    );
+    const artifactNode: ElementNode | undefined = dependencyNode?.children?.find(node => node.tag === XmlTagName.ArtifactId);
+    if (artifactNode !== undefined) {
+        await locateInFile(pomPath, artifactNode);
+    } else {
+        throw new Error("Open the wrong pom file and can not find where the dependency has been imported.");
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:43]: grammar problem and confusing message.

> Open the wrong pom file and

Are you sure the pom file is wrong? Can it be caused by wrong artifact id?


Suggest to simply use "Failed to locate the dependency."

File: src/handlers/jumpToDefinitionHandler.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        throw new UserError("Invalid target XML node to locate.");
+    }
+    const currentDocument: vscode.TextDocument = await vscode.workspace.openTextDocument(pomPath);
+    const textEditor: vscode.TextEditor = await vscode.window.showTextDocument(currentDocument, {preview: false});
+    const start = currentDocument.positionAt(targetNode.contentStart);
+    textEditor.selection = new vscode.Selection(start, start);
+    textEditor.revealRange(new vscode.Range(start, start), vscode.TextEditorRevealType.InCenter);
+}
+
+function localPomPath(gid: string, aid: string, version: string): string {
```
Comment (Eskibear) [inline src/handlers/jumpToDefinitionHandler.ts:58]: I remember similar utility is also used in code completion. It would be nice to extract.

**Review by Eskibear — CHANGES_REQUESTED** [review]
1. rename all `jumpToDefinition` to `goTo...` for consistency with vscode.
2. error messages
3. problems in finding position of target dependency

File: package.json (hunk 3 above)
Comment (Melody618) [reply package.json:61]: What about `excludeDependnecy` and `setVersionto`? I think they are also unnecessary.

Comment (Eskibear) [reply package.json]: it depends on whether it's available in command palette. If not, you should remove it.

**Review by Eskibear — APPROVED** [review]

**Pull request merged**